velox/functions/prestosql/aggregates/PrestoHasher.cpp

```diff
@@ -0,0 +1,208 @@
+/*
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#define XXH_INLINE_ALL
+
+#include "velox/functions/prestosql/aggregates/PrestoHasher.h"
+#include "velox/external/xxhash.h"
+
+namespace facebook::velox::aggregate {
+
+namespace {
+
+int64_t rotateLeft(const int64_t num, const int32_t distance) {
+  auto unsignedShift = ((uint64_t)num) >> -distance;
+  return (num << distance) | unsignedShift;
+}
+
+template <typename T>
+FOLLY_ALWAYS_INLINE int64_t getHashForIntTypes(const T& value) {
+  return rotateLeft(value * 0xC2B2AE3D27D4EB4FL, 31) * 0x9E3779B185EBCA87L;
+}
+
+template <typename T>
+FOLLY_ALWAYS_INLINE int64_t
+getHashForIntTypes(const DecodedVector& vector, vector_size_t row) {
+  auto value = vector.valueAt<T>(row);
+  return getHashForIntTypes(value);
+}
+
+FOLLY_ALWAYS_INLINE int64_t
+getHashForVarCharTypes(const DecodedVector& vector, vector_size_t row) {
+  auto input = vector.valueAt<StringView>(row);
+  int64_t hash = XXH64(input.data(), input.size(), 0);
+  return hash;
+}
+} // namespace
+
+template <>
+int64_t PrestoHasher::computeHash<TypeKind::BIGINT>(vector_size_t row) {
+  return getHashForIntTypes<int64_t>(*vector_.get(), row);
+}
+
+template <>
+int64_t PrestoHasher::computeHash<TypeKind::INTEGER>(vector_size_t row) {
+  return getHashForIntTypes<int32_t>(*vector_.get(), row);
+}
+
+template <>
+int64_t PrestoHasher::computeHash<TypeKind::SMALLINT>(vector_size_t row) {
+  return getHashForIntTypes<int16_t>(*vector_.get(), row);
+}
+
+template <>
+int64_t PrestoHasher::computeHash<TypeKind::TINYINT>(vector_size_t row) {
+  return getHashForIntTypes<int8_t>(*vector_.get(), row);
+}
+
+template <>
+int64_t PrestoHasher::computeHash<TypeKind::DOUBLE>(vector_size_t row) {
+  auto value = vector_->valueAt<double>(row);
+  int64_t u = *reinterpret_cast<int64_t*>(&value);
```

Comment:
- DecodedVector::valueAt<T> already does reinterpret_cast, hence, you can simplify this code to 
 
`vector->valueAt<int64_t>(row)`

- Renaming getHashForIntTypes and getHashForVarCharTypes to hashInteger and hashStringView might improve readability:

```
return hashInteger(vector->valueAt<int64_t>(row));
...
return hashStringView(vector->valueAt<StringView>(row));
```